`//ros2/test/rosbag/...` tests are flaky

`❯ bazel test --runs_per_test 100 //ros2/test/rosbag/...`
```
//ros2/test/rosbag:mcap_tests        FAILED in 10 out of 100 in 2.9s
//ros2/test/rosbag:sqlite3_tests     FAILED in 4 out of 100 in 5.1s
```

Example log:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //ros2/test/rosbag:mcap_tests
-----------------------------------------------------------------------------
Running with ROS_DOMAIN_ID 1
[INFO] [launch]: All log files can be found below /home/laurenz/.cache/bazel/_bazel_laurenz/560f8735b75dc5afa81db7c5a8d429bc/sandbox/linux-sandbox/1040/execroot/com_github_mvukov_rules_ros2/bazel-out/k8-fastbuild/testlogs/ros2/test/rosbag/mcap_tests/run_99_of_100/test.outputs/2023-03-08-21-08-18-698721-tpxy-12
[INFO] [launch]: Default logging verbosity is set to INFO
test_recorder (tests.TestRecorder) ... [INFO] [publisher-1]: process started with pid [17]
[INFO] [recorder-2]: process started with pid [19]
[publisher-1] 1678309698.750218 [1]  publisher: selected interface "lo" is not multicast-capable: disabling multicast
[recorder-2] 1678309698.765050 [1]   recorder: selected interface "lo" is not multicast-capable: disabling multicast
[recorder-2] stdin is not a terminal device. Keyboard handling disabled.[INFO] [1678309698.782350883] [rosbag2_recorder]: Press SPACE for pausing/resuming
[recorder-2] [INFO] [1678309698.788707353] [rosbag2_recorder]: Listening for topics...
[recorder-2] [INFO] [1678309698.788757080] [rosbag2_recorder]: Event publisher thread: Starting
[recorder-2] [INFO] [1678309698.891732503] [rosbag2_recorder]: Subscribed to topic '/topic'
[recorder-2] [INFO] [1678309698.891949565] [rosbag2_recorder]: All requested topics are subscribed. Stopping discovery...
[recorder-2] [INFO] [1678309698.989649000] [rosbag2_recorder]: Event publisher thread: Exiting
[INFO] [recorder-2]: process has finished cleanly [pid 19]
[INFO] [publisher-1]: sending signal 'SIGINT' to process[publisher-1]
[publisher-1] [INFO] [1678309699.237376828] [rclcpp]: signal_handler(signum=2)
[INFO] [publisher-1]: process has finished cleanly [pid 17]
FAIL

======================================================================
FAIL: test_recorder (tests.TestRecorder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/laurenz/.cache/bazel/_bazel_laurenz/560f8735b75dc5afa81db7c5a8d429bc/sandbox/linux-sandbox/1040/execroot/com_github_mvukov_rules_ros2/bazel-out/k8-fastbuild/bin/ros2/test/rosbag/mcap_tests.runfiles/com_github_mvukov_rules_ros2/ros2/test/rosbag/tests.py", line 77, in test_recorder
    self.assertGreaterEqual(bag_metadata['message_count'],
AssertionError: 4 not greater than or equal to 9

----------------------------------------------------------------------
Ran 1 test in 0.486s

FAILED (failures=1)

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
```

--- a/ros2/test/rosbag/publisher.cc
+++ b/ros2/test/rosbag/publisher.cc
@@ -24,16 +24,20 @@ class Publisher : public rclcpp::Node {
     publisher_ = create_publisher<std_msgs::msg::String>("topic", 10);
     auto timer_callback = [this]() -> void {
       auto message = std_msgs::msg::String();
-      message.data = std::to_string(count_++);
+      // Produce reasonably large messages so we quickly reach the size limit
+      // that causes the recorder to split the bag (we use this as a signal that
+      // the recorder has received a sufficient number of messages).
+      constexpr std::size_t minimum_message_size = 8192;
+      constexpr char random_char = 'x';
+      message.data = std::string(minimum_message_size, random_char);
       publisher_->publish(message);
     };
     timer_ = create_wall_timer(std::chrono::milliseconds(20), timer_callback);
   }
 
  private:
-  rclcpp::TimerBase::SharedPtr timer_;
   rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
-  size_t count_ = 0;
+  rclcpp::TimerBase::SharedPtr timer_;
 };
 
 int main(int argc, char* argv[]) {

--- a/ros2/test/rosbag/publisher.cc
+++ b/ros2/test/rosbag/publisher.cc
@@ -27,9 +27,8 @@ class Publisher : public rclcpp::Node {
       // Produce reasonably large messages so we quickly reach the size limit
       // that causes the recorder to split the bag (we use this as a signal that
       // the recorder has received a sufficient number of messages).
-      constexpr std::size_t minimum_message_size = 8192;
-      constexpr char random_char = 'x';
-      message.data = std::string(minimum_message_size, random_char);
+      constexpr std::size_t kMinimumMessageSize = 8192;
+      message.data = std::string(kMinimumMessageSize, 'x');
       publisher_->publish(message);
     };
     timer_ = create_wall_timer(std::chrono::milliseconds(20), timer_callback);

--- a/ros2/test/rosbag/recorder.cc
+++ b/ros2/test/rosbag/recorder.cc
@@ -13,6 +13,7 @@
 // limitations under the License.
 #include "rosbag2_transport/recorder.hpp"
 #include "rclcpp/rclcpp.hpp"
+#include "rosbag2_interfaces/msg/write_split_event.hpp"
 #include "rosbag2_storage/storage_options.hpp"
 #include "rosbag2_transport/reader_writer_factory.hpp"
 #include "rosbag2_transport/record_options.hpp"
@@ -20,27 +21,27 @@
 
 constexpr auto kTopicName = "topic";
 
-class MessageCounter : public rclcpp::Node {
+class BagSplitEventListener : public rclcpp::Node {
  public:
-  MessageCounter() : Node("message_counter") {
-    subscription_ = create_subscription<std_msgs::msg::String>(
-        kTopicName, 10,
-        [this](std_msgs::msg::String::SharedPtr /*msg*/) { ++msg_count_; });
+  using MessageT = rosbag2_interfaces::msg::WriteSplitEvent;
+
+  BagSplitEventListener() : Node("bag_split_event_listener") {
+    subscription_ = create_subscription<MessageT>(
+        "events/write_split", 10,
+        [this](MessageT::SharedPtr /*msg*/) { ++split_count_; });
   }
 
-  auto msg_count() const { return msg_count_; }
+  auto split_count() const { return split_count_; }
 
  private:
-  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
-  int msg_count_ = 0;
+  rclcpp::Subscription<MessageT>::SharedPtr subscription_;
+  int split_count_ = 0;
 };
 
 int main(int argc, char* argv[]) {
   rclcpp::init(argc, argv);
 
   rclcpp::executors::SingleThreadedExecutor executor;
-  auto msg_counter = std::make_shared<MessageCounter>();
-  executor.add_node(msg_counter);
 
   rosbag2_transport::RecordOptions record_options;
   record_options.topics = {kTopicName};
@@ -54,19 +55,22 @@ int main(int argc, char* argv[]) {
       rosbag2_transport::ReaderWriterFactory::make_writer(record_options);
 
   rosbag2_storage::StorageOptions storage_options;
-  storage_options.uri =
-      std::string(std::getenv("TEST_UNDECLARED_OUTPUTS_DIR")) + "/bag";
+  storage_options.uri = std::string(std::getenv("TEST_TMPDIR")) + "/bag";
   storage_options.storage_id = std::getenv("STORAGE_ID");
+  // we use the bag split event as a signal that we can stop
+  constexpr uint64_t sqlite3_minimum_split_size = 86016;
+  storage_options.max_bagfile_size = sqlite3_minimum_split_size;
   auto recorder = std::make_shared<rosbag2_transport::Recorder>(
       std::move(writer), storage_options, record_options);
-  recorder->record();
   executor.add_node(recorder);
 
-  while (rclcpp::ok()) {
+  auto split_counter = std::make_shared<BagSplitEventListener>();
+  executor.add_node(split_counter);
+
+  recorder->record();
+
+  while (rclcpp::ok() && split_counter->split_count() == 0) {
     executor.spin_once(std::chrono::milliseconds(10));
-    if (msg_counter->msg_count() > 10) {
-      break;
-    }
   }
 
   return rclcpp::shutdown() ? EXIT_SUCCESS : EXIT_FAILURE;

--- a/ros2/test/rosbag/recorder.cc
+++ b/ros2/test/rosbag/recorder.cc
@@ -23,18 +23,18 @@ constexpr auto kTopicName = "topic";
 
 class BagSplitEventListener : public rclcpp::Node {
  public:
-  using MessageT = rosbag2_interfaces::msg::WriteSplitEvent;
+  using WriteSplitEvent = rosbag2_interfaces::msg::WriteSplitEvent;
 
   BagSplitEventListener() : Node("bag_split_event_listener") {
-    subscription_ = create_subscription<MessageT>(
+    subscription_ = create_subscription<WriteSplitEvent>(
         "events/write_split", 10,
-        [this](MessageT::SharedPtr /*msg*/) { ++split_count_; });
+        [this](WriteSplitEvent::SharedPtr /*msg*/) { ++split_count_; });
   }
 
   auto split_count() const { return split_count_; }
 
  private:
-  rclcpp::Subscription<MessageT>::SharedPtr subscription_;
+  rclcpp::Subscription<WriteSplitEvent>::SharedPtr subscription_;
   int split_count_ = 0;
 };
 
@@ -57,9 +57,9 @@ int main(int argc, char* argv[]) {
   rosbag2_storage::StorageOptions storage_options;
   storage_options.uri = std::string(std::getenv("TEST_TMPDIR")) + "/bag";
   storage_options.storage_id = std::getenv("STORAGE_ID");
-  // we use the bag split event as a signal that we can stop
-  constexpr uint64_t sqlite3_minimum_split_size = 86016;
-  storage_options.max_bagfile_size = sqlite3_minimum_split_size;
+  // We use the bag split event as a signal that we can stop.
+  constexpr uint64_t kSqlite3MinimumSplitSize = 86016;
+  storage_options.max_bagfile_size = kSqlite3MinimumSplitSize;
   auto recorder = std::make_shared<rosbag2_transport::Recorder>(
       std::move(writer), storage_options, record_options);
   executor.add_node(recorder);

--- a/ros2/test/rosbag/tests.py
+++ b/ros2/test/rosbag/tests.py
@@ -61,7 +61,7 @@ class TestRecorder(unittest.TestCase):
             proc_info,
             allowable_exit_codes=[launch_testing.asserts.EXIT_OK],
             process=recorder)
-        bag_dir = os.path.join(os.environ['TEST_UNDECLARED_OUTPUTS_DIR'], 'bag')
+        bag_dir = os.path.join(os.environ['TEST_TMPDIR'], 'bag')
         bag_db_file = os.path.join(bag_dir,
                                    STORAGE_IDS_TO_BAG_NAMES[get_storage_id()])
         bag_metadata_file = os.path.join(bag_dir, 'metadata.yaml')

